File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        d={}
        start=0
        end=0
        maxAnswer=0
        while end<len(s):
            if s[end] not in d:
                d[s[end]]=d.get(s[end],0)+1
                end=end+1
                maxAnswer=max(maxAnswer,len(d))
            else:
                while s[end] in d and start<=end:
                    d[s[start]]=d.get(s[start],0)-1
                    if d[s[start]]==0:
                        del d[s[start]]
                    start=start+1
                d[s[end]]=d.get(s[end],0)+1
                end=end+1
                maxAnswer=max(maxAnswer,len(d))
        return maxAnswer


        

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int length=s.length();
        int left=0;
        int right=0;
        int []hash=new int[256];
        Arrays.fill(hash,-1);
        int maxi=0;
        while(right<length){
            if(hash[s.charAt(right)]!=-1){
                left=Math.max(left,hash[s.charAt(right)]+1);
            }
            maxi=Math.max(maxi,right-left+1);
            hash[s.charAt(right)]=right;
            right++;
        }
        return maxi;
    }
}